src/Lib.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Lib (app) where
+
+import Control.Monad
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Except (ExceptT (..), runExceptT, withExceptT)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import Data.Coerce (coerce)
+import Data.Either (isLeft)
+import Data.Maybe (fromMaybe)
+import Database.SQLite.Simple
+import Database.SQLite.Simple.FromField (FromField (..))
+import Database.SQLite.Simple.ToField (ToField (..))
+import GHC.IO.Exception (ExitCode (ExitSuccess))
+import Network.Wai.Handler.Warp (run)
+import Servant
+import qualified System.Directory as Dir
+import System.FilePath ((</>))
+import qualified System.Process as Proc
+
+-- | Top-level API. Examples:
+--   localhost:8081/ps
+--   localhost:8081/procps/ps?sys=aarch64-linux
+--  TODO API documentation
+type API =
+  Get '[JSON] NoContent
+    :<|> Capture "binary-name" BinaryName :> QueryParam "sys" System :> Get '[OctetStream] ByteString
+    :<|> Capture "package-name" PackageName :> Capture "binary-name" BinaryName :> QueryParam "sys" System :> Get '[OctetStream] ByteString
+
+newtype BinaryName = BinaryName {unBinaryName :: String}
+  deriving newtype (ToField, FromHttpApiData)
+
+newtype PackageName = PackageName {unPackageName :: String}
+  deriving newtype (Eq, Ord, FromField, FromHttpApiData, ToField)
+
+-- TODO just make this a string?
```

Comment:
👍 for sum type rather than a string.

Reply:
I think I agree. Although it does mean we have to hand-code every system we support.

Reply:
Oh, also, how do you think we should store them in the DB?

1. as strings
2. as an integer enum
2. as foreign keys into a `systems` table

I feel like there's something to be said for each of these?

Reply:
Storing it in the DB is an idea.
In that case, "1. as strings" + unique constraint looks good enough.

Reply:
Oh, I meant, "what is the type of the `system` column in the `binary` table". If I understand correctly, what you're describing is situation 3 in the list above. Do you think that's better than just having it be text (situation 1.) as is currently the case?

https://github.com/binplz/binplz.dev/blob/53cb018cdfab40b86aec26a60a1dca45bd54c5bf/src/Lib.hs#L180